src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,472 @@
+                }
+
+                Debug.Assert(index == _preCount);
+                foreach (TSource item in _source)
+                {
+                    array[index] = item;
+                    ++index;
+                }
+
+                Debug.Assert(index == array.Length - _appCount);
+                for (int i = 0; i != _appCount; ++i, ++index)
+                {
+                    array[index] = _appended[i];
+                }
+
+                Debug.Assert(index == array.Length);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(Math.Max(_appCount + _preCount, 4)) : new List<TSource>(count);
+                for (int i = -_preCount; i != 0; ++i)
+                {
+                    list.Add(_prepended[i]);
+                }
+
+                list.AddRange(_source);
+
+                for (int i = 0; i != _appCount; ++i)
+                {
+                    list.Add(_appended[i]);
+                }
+
+                return list;
+            }
+        }
+    }
+
+    internal class SharedCollection<T>
+    {
+        // Limit the number of elements stored, because the collection being shared makes it possible
+        // for a series of operations to grow the collection large, but for only one of the earlier
+        // operations to be keeping it alive.
+        private const int MaxLength = 64;
+        private T[] _store;
+        private int _count;
+
+        public SharedCollection(T item)
+        {
+            _store = new[] { item };
+            _count = 1;
+        }
+
+        public SharedCollection(T first, T second)
+        {
+            _store = new[] { first, second };
+            _count = 2;
+        }
+
+        public T this[int index] => _store[index >= 0 ? index : index + _store.Length];
+
+        public bool Append(T item, int number)
+        {
+            if (number != _count | number >= MaxLength)
+            {
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+                T[] newStore = new T[number << 1];
+                for (int i = 0; i != store.Length; ++i)
+                {
+                    newStore[i] = store[i];
+                }
+
+                _store = store = newStore;
+            }
+
+            store[number] = item;
+            ++_count;
+            return true;
+        }
+
+        public bool Prepend(T item, int number)
+        {
+            if (number != _count | number >= MaxLength)
+            {
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+
+                T[] newStore = new T[number << 1];
+                for (int i = 0; i != store.Length; ++i)
```

Comment:
I very much doubt it. Preference from a long time ago being more often using C++ iterator objects where `!=` might be defined for the type but `<` might not. (And longer ago again it might perhaps have made for a slight but measurable performance difference on old CPUs).